### PR TITLE
Refactor svg read write

### PR DESCRIPTION
**Pull request by rastogi-bhavya**
Changes in accordance to #38 

1. Added default color (red): This is done to make sure that in case there is an SVG uploaded by the user which does not have `stroke` defined, then at least software will be able to show the SVG components.

2. Read Corrections: Changes were made to how the software read an SVG.
- When the object is created, the set of attributes are initialized by their default value.
- When the SVG object is read `this._attrs` is updated with the new values.
- For `style`, the individual properties are also added as separate values, to make updates easier. (the style is also added to `this._attrs`)

3. Displaying an SVG: When displaying the SVG, now the `style` attribute is no longer used, as all of its properties are already in `'this._attrs`

4. Write Corrections: When writing attributes back, a simple which is made to see if that attribute is present in `style`, if yes then that property in style is updated, else the attribute is written back. The `style` is added at the end, after all other attributes have been written.

Added an annotation-utils file for re-usability of code.

### CODE REVIEW

File: js/viewer/annotation/annotation-svg.js [hunk 1]
```diff
@@ -528,7 +499,7 @@ function AnnotationSVG(parent, id, imgWidth, imgHeight, scale, ignoreReferencePo
                     if(id !== "undefined"){
                         group = this.groups.hasOwnProperty(id) ? this.groups[id] : this.createAnnotationGroup(id, anatomy);
                         annotation = group.addAnnotation(node.nodeName);
-                        annotation.setAttributesBySVG(node);
+                        annotation.setAttributesByJSON(this.getNodeAttributes(node.attributes));
```
Comment (RFSH) [inline js/viewer/annotation/annotation-svg.js:502]: Instead of passing `node.attributes`, pass `node`.

File: js/viewer/annotation/annotation-svg.js [hunk 2]
```diff
@@ -488,6 +446,14 @@ function AnnotationSVG(parent, id, imgWidth, imgHeight, scale, ignoreReferencePo
         return node.getAttribute("id") || node.getAttribute("name") || parentNode.getAttribute("id") || parent.all_config.constants.UNKNOWN_ANNNOTATION;
     }
 
+    this.getNodeAttributes = function(node) {
```
Comment (RFSH) [inline js/viewer/annotation/annotation-svg.js:449]: Add jsdoc to this function

File: js/app.js [hunk 3]
```diff
@@ -169,6 +169,9 @@ var myApp = (function (_config) {
                 case 'saveAnnotationRecord':
                     viewer.saveAnatomySVG(data);
                     break;
+                // case 'discardAnnotationChanges':
```
Comment (RFSH) [inline js/app.js:172]: remove this

File: js/config.js [hunk 4]
```diff
@@ -74,7 +74,9 @@ var _config = {
...
           OUTPUT_SVG_STROKE_WIDTH: '1px',
           DISABLE_VECTOR_EFFECT: false,
-          DEFAULT_LINE_THICKNESS: '1px'
+          DEFAULT_LINE_THICKNESS: '1px',
+          DEFAULT_STROKE: 'red',
+          SVG_ATTRIBUTE_LIST: ["fill", "stroke", "stroke-width", "vector-effect", "cx", "cy", "r", "d", "x", "y", "height", "width"]
```
Comment (RFSH) [inline js/config.js:79]: What's the point of this?

File: js/viewer/annotation/base.js [hunk 5]
```diff
@@ -6,16 +6,25 @@ function Base(attrs){
     this.isDrawing = false;
     this.constants = attrs.parent.parent.parent.all_config.constants;
     this.urlParams = attrs.parent.parent.parent.parameters;
+    this.annotationUtils = new AnnotationUtils();
```
Comment (RFSH) [inline js/viewer/annotation/base.js:9]: Since you moved all the logic to be in the annotation-svg, you don't need the utils here anymore.

File: js/viewer/annotation/base.js [hunk 6]
```diff
@@ -27,27 +36,6 @@ function Base(attrs){
...
-        }
-
-        if (styleString[styleString.length - 1] == ';') {
-            styleString = styleString.slice(0, styleString.length - 1);
-        }
-
-        return styleString;
-    }
 
     this.exportToSVG = function(){
```
Comment (RFSH) [inline js/viewer/annotation/base.js:40]: This function is still incomplete.
- You're still iterating the objects without any particular order. So your output won't be deterministic. 
- You're not properly handling the `addedProperties`.

File: js/viewer/annotation/base.js [hunk 7]
```diff
@@ -6,16 +6,25 @@ function Base(attrs){
     this.isDrawing = false;
     this.constants = attrs.parent.parent.parent.all_config.constants;
     this.urlParams = attrs.parent.parent.parent.parameters;
+    this.annotationUtils = new AnnotationUtils();
+
+    // this is used to make sure that any property that is added to the SVG for display purpose, is not added while saving
+    this.addedProperties = {
+        "vector-effect": true
+    };
     this._attrs = {
```
Comment (RFSH) [inline js/viewer/annotation/base.js:15]: Why did you remove the parts that we were getting the values from `attrs`? Based on what I'm seeing in `attribute-group.js` we're sending `graph-id`. Do you even know why we have `graph-id` as an attribute?

File: js/viewer/annotation/base.js [hunk 8]
```diff
@@ -6,16 +6,25 @@ function Base(attrs){
     this.isDrawing = false;
     this.constants = attrs.parent.parent.parent.all_config.constants;
     this.urlParams = attrs.parent.parent.parent.parameters;
+    this.annotationUtils = new AnnotationUtils();
+
+    // this is used to make sure that any property that is added to the SVG for display purpose, is not added while saving
+    this.addedProperties = {
```
Comment (RFSH) [inline js/viewer/annotation/base.js:12]: In pure Javascript, there's no concept of private/public APIs. So instead we use a naming convention for this. If an attribute starts with `_`, it's private, otherwise, it will be considered public. So changed `addedProperties` and `ignoredAttributes` to `_addedProperties` and `_ignoredAttributes`.

File: js/viewer/annotation/base.js [hunk 9]
```diff
@@ -132,6 +110,18 @@ function Base(attrs){
     };
 }
 
+/**
+ * This functions check to see if the properties mentioned in this.addedProperties are present in the attribute list which is used to initialize this object.
+ * @param {object} attrs
+ */
+Base.prototype.updateAddedProperties = function (attrs) {
```
Comment (RFSH) [inline js/viewer/annotation/base.js:117]: You don't need this function, you can do it directly in the `setAttributesByJSON` without the need to iterate over the `attrs` again. So

```javascript
Base.prototype.setAttributesByJSON = function (attrs) {
  var attr,
        value,
        discardedAttributes = ["id", "style"];

  for (attr in attrs) {
    if (!attr || !attrs.hasOwnProperty(attr) || discaredAttributes.indexOf(attr) === -1) continue;
    value = attrs[attr];
    
    if (attr in this._addedProperties) {
      this._addedProperties[attr] = value;
    } else if (attr in this._attrs) {
      this._attrs[attr] = value;
    } else {
      this._ignoredAttributes[attr] = value;
    }
       
  
  }
};
```

File: js/viewer/annotation/base.js [hunk 10]
```diff
@@ -60,36 +48,26 @@ function Base(attrs){
...
-
         for(attr in this._attrs){
-            // Ingoring style because that will be added at the end.
-            // Ignoreing vector-effect becasue this properpty is not needed in the output file
-            if (!this._attrs[attr] || this._attrs[attr] === null || attr === "style" || attr == "vector-effect"){
+            if (!this._attrs[attr] || this._attrs[attr] === null){
                 continue;
             }
 
             switch(attr){
```
Comment (RFSH) [inline js/viewer/annotation/base.js:56]: You should figure out what `tag` and `graph-id` are. I don't think we need them as attributes. Also you shouldn't just add `vector-effect` and instead should go over the `addedProperties` list. So if there's an attribute in `_attrs` that is part of `addedProperties`, we shouldn't add it to export and instead use the value that comes from `addedProperties`.

File: js/viewer/annotation/annotation-svg.js (hunk 1 above)
Comment (rastogi-bhavya) [reply js/viewer/annotation/annotation-svg.js:502]: Why should `node` be passed instead of `node.attribute` ?

File: js/app.js (hunk 3 above)
Comment (rastogi-bhavya) [reply js/app.js:172]: So this should be added when I do the commit for the Undo issue?

File: js/config.js (hunk 4 above)
Comment (rastogi-bhavya) [reply js/config.js:79]: this was added so that i have a list of attributes that are being handled, and the order in which they should be present in the output. But since we have also have to add the `ignored-attributes`, I thought it would be better to just sort all the attributes in ascending order, and therefore ended up not using them. Should I remove `SVG_ATTRIBUTE_LIST`?

File: js/viewer/annotation/base.js (hunk 6 above)
Comment (rastogi-bhavya) [reply js/viewer/annotation/base.js:40]: This has  been added.

File: js/viewer/annotation/annotation-svg.js (hunk 1 above)
Comment (RFSH) [reply js/viewer/annotation/annotation-svg.js:502]: It's called `getNodeAttributes`, so you'd expect to get it from the node. Also, your function implementation is just confusing and not easy to look at as it's assuming the input. If your input is an `Element` typed object it's more straightforward.

Comment (rastogi-bhavya) [reply js/viewer/annotation/annotation-svg.js:502]: yeah that makes sense. I will correct that. Are you talking about the implementation of the `getNodeAttributes` function. If so then that is because the `node.attributes` is a JS `NamedNodeMap` (https://developer.mozilla.org/en-US/docs/Web/API/NamedNodeMap), and therefore i had to make the function so that I can get the desired output.

File: js/viewer/annotation/base.js (hunk 9 above)
Comment (rastogi-bhavya) [reply js/viewer/annotation/base.js:129]: The reason I had this function was because, earlier I had the function `setAttributesBySVG` and I needed to update the `_addedProperties` in that too, so instead of doing the same thing twice, I created function for it.
The other reason was to emphasize on the need for that processing. Having a separate function provides that emphasis.

Should I remove it?

File: js/viewer/annotation/base.js (hunk 10 above)
Comment (rastogi-bhavya) [reply js/viewer/annotation/base.js]: `tag` = {rect, circle, line}
I still have to figure out why `graph-id` is used. From what I understand it is used for delete and highlight functionality.
I have added the check for `_addedProperties`. https://github.com/informatics-isi-edu/openseadragon-viewer/blob/4c4eb19b816508fafca08b8782afab2cf4796e76/js/viewer/annotation/base.js#L62

File: js/viewer/annotation/base.js (hunk 9 above)
Comment (RFSH) [reply js/viewer/annotation/base.js:129]: If it's just thing one place then yes. Right now, you're iterating over the list of attributes twice. If you do it like I suggested you're going over the attribute list only once.

File: js/app.js (hunk 3 above)
Comment (RFSH) [reply js/app.js:172]: Yup.

File: js/viewer/annotation/base.js (hunk 7 above)
Comment (rastogi-bhavya) [reply js/viewer/annotation/base.js:15]: I removed that because, but initialization those value were not passed, and were set later during the `set` functions. What I did do is initialize them with default values.

I dont when why we are sending the `graph-id` or why we use it in attributes

File: js/viewer/annotation/annotation-svg.js (hunk 2 above)
Comment (rastogi-bhavya) [reply js/viewer/annotation/annotation-svg.js:454]: Done

File: js/app.js (hunk 3 above)
Comment (rastogi-bhavya) [reply js/app.js]: Removed

File: js/viewer/annotation/base.js (hunk 5 above)
Comment (rastogi-bhavya) [reply js/viewer/annotation/base.js]: Removed

File: js/viewer/annotation/base.js (hunk 8 above)
Comment (rastogi-bhavya) [reply js/viewer/annotation/base.js]: Done

File: js/viewer/annotation/base.js (hunk 7 above)
Comment (RFSH) [reply js/viewer/annotation/base.js:14]: You should figure it out. With what you did here, you're not using the `graph-id` value that is passed from the `attribute-group.js` (`addAnnotation` function).

If `graph-id` is not needed, you should modify `Path`, `Polyline`, `Polygon`, `Rect`, `Circle`, and `Base` constructor functions to only accept a `parent` (instead of `attrs`).

File: js/viewer/annotation/base.js [hunk 11]
```diff
@@ -81,15 +58,27 @@ function Base(attrs){
                 case "graph-id":
```
Comment (RFSH) [inline js/viewer/annotation/base.js:58]: You should figure out why there's `tag` and `graph-id` in attributes. If these are values that you need in the object, they shouldn't be part of `_attrs` and instead should be separate properties of your class.

File: js/viewer/annotation/path.js [hunk 12]
```diff
@@ -5,10 +5,6 @@ function Path(attrs){
 
     this._attrs["tag"] = "path";
     this._attrs["d"] = attrs["d"] || "";
-    this._attrs["fill"] = attrs["fill"] || "";
```
Comment (RFSH) [inline js/viewer/annotation/path.js:8]: Why did you remove these? and why didn't you remove them from the other prototypes (other shapes)?

File: js/config.js (hunk 4 above)
Comment (RFSH) [reply js/config.js:79]: Yes, you should remove any piece of code that you add along the way and never use.

File: js/viewer/annotation/path.js (hunk 12 above)
Comment (rastogi-bhavya) [reply js/viewer/annotation/path.js:8]: I removed `fill` from path because it is already present in `base.js`. Fill is not there in `circle` and `rect`, and if you were talking about 'polyline` and `polygon` then, they are just empty container files for future use, so I did not think necessary to make changes in then. I can do so if you want to.

Comment (RFSH) [reply js/viewer/annotation/path.js:8]: Yes, that's what I'm talking about. Remove it from the other prototypes as well. It doesn't matter that they're not currently used, if you don't change the code consistently, we won't remember why we did it that way.

File: js/viewer/annotation/base.js (hunk 11 above)
Comment (RFSH) [reply js/viewer/annotation/base.js:58]: If you plan on leaving the `graph-id`, you should at least get rid of this useless switch case. You can just add `attr === "tag" || attr === "graph-id"` to the if statement above.

File: js/viewer/annotation/base.js (hunk 9 above)
Comment (rastogi-bhavya) [reply js/viewer/annotation/base.js]: Done

File: js/config.js (hunk 4 above)
Comment (rastogi-bhavya) [reply js/config.js]: Done

File: js/viewer/annotation/path.js (hunk 12 above)
Comment (rastogi-bhavya) [reply js/viewer/annotation/path.js:8]: Okay, so I will also remove other things that are redundant or not being used.

File: js/viewer/annotation/base.js (hunk 11 above)
Comment (rastogi-bhavya) [reply js/viewer/annotation/base.js]: Done

File: js/viewer/annotation/base.js (hunk 7 above)
Comment (rastogi-bhavya) [reply js/viewer/annotation/base.js:17]: I have added a TODO in the code for the above requirement.

File: js/viewer/annotation/base.js [hunk 13]
```diff
@@ -27,69 +39,42 @@ function Base(attrs){
...
+            attributeList[attrs] = this._ignoredAttributes[attr];
         }
+
+        // sort the attributes to make the output deterministic
+        var sortedAttributeList = {};
+        Object.keys(attributeList).sort().forEach(function (key) {
+            sortedAttributeList[key] = attributeList[key];
+        });
+
+        for(attr in sortedAttributeList){
```
Comment (RFSH) [inline js/viewer/annotation/base.js:74]: Iterating over objects doesn't necessarily preserve the order. You have the keys sorted up there, so add to output directly up there. You don't need to create another object, so

```javascript 
Object.keys(attributeList).sort().forEach(function (key) {
  rst += (key + '="' + attributeList[key] + '" ');
});
```

Comment (rastogi-bhavya) [reply js/viewer/annotation/base.js]: Done
